sdk/storage/MigrationGuide.md

```diff
@@ -0,0 +1,477 @@
+![Blob Storage Hierarchy](https://docs.microsoft.com/en-us/azure/storage/blobs/media/storage-blobs-introduction/blob1.png)
+
+In the interest of simplifying the API surface, v12 uses three top level clients to match this structure that can be used to interact with a majority of your resources: `BlobServiceClient`, `BlobContainerClient`, and `BlobClient`. Note that blob-type-specific operations can still be accessed by their specific clients, as in v7.5.
+
+#### Migrating from cloud_blob_directory
+
+Note the absence of a v12 equivalent for v7.5's `cloud_blob_directory`. Directories were an SDK-only concept that did not exist in Azure Blob Storage, and which were not brought forwards into the modern Storage SDK. As shown by the diagram in [Client Structure](#client-structure), containers only contain a flat list of blobs, but those blobs can be named and listed in ways that imply a folder-like structure. See our [Listing Blobs in a Container](#listing-blobs-in-a-container) migration samples later in this guide for more information.
+
+For those whose workloads revolve around manipulating directories and heavily relied on the leagacy SDKs abstraction of this structure, consider the [pros and cons of enabling hierarchical namespace](https://docs.microsoft.com/azure/storage/blobs/data-lake-storage-namespace) on your storage account, which would allow switching to the [Data Lake Gen 2 SDK](https://github.com/Azure/azure-sdk-for-cpp/tree/master/sdk/storage/azure-storage-files-datalake), whose migration is not covered in this document.
+
+#### Class Conversion Reference
+
+The following table lists v7.5 classes and their v12 equivalents for quick reference.
+
+| v7.5 | v12 |
+|-------|--------|
+| `cloud_blob_client` | `BlobServiceClient` |
+| `cloud_blob_container`  | `BlobContainerClient` |
+| `cloud_blob_directory` | No equivalent |
+| `cloud_blob` | `BlobClient` |
+| `cloud_block_blob` | `BlockBlobClient` |
+| `cloud_page_blob` | `PageBlobClient` |
+| `cloud_append_blob` | `AppendBlobClient` |
+
+## Migration Samples
+
+### Creating a Container
+
+v7.5
+```C++
+auto container_client = service_client.get_container_reference(container_name);
+container_client.create();
+```
+
+v12
+```C++
+auto containerClient = serviceClient.GetBlobContainerClient(containerName);
+containerClient.Create();
+```
+
+Or you can use the `BlobServiceClient.CreateBlobContainer()` method.
+
+```C++
+serviceClient.CreateBlobContainer(containerName);
+```
+
+### Uploading Blobs to a Container
+
+#### Uploading from a file
+
+v7.5
+```C++
+cloud_block_blob block_blob_client = container_client.get_block_blob_reference(blob_name);
+block_blob_client.upload_from_file(local_file_path);
+```
+
+v12
+```C++
+BlockBlobClient blockBlobClient = containerClient.GetBlockBlobClient(blobName);
+blockBlobClient.UploadFrom(localFilePath);
+```
+
+#### Uploading from a stream
+
+v7.5
+```C++
+block_blob_client.upload_from_stream(stream);
+```
+
+v12
+```C++
+blockBlobClinet.Upload(stream);
```

Comment:
Is this single threaded? I thought the equivalent should be `UploadFrom`?

Reply:
`Upload` uploads from stream
`UploadFrom` uploads from file or memory buffer